README
AnyEvent-Redis2

AnyEvent::Redis2 is an event-driven (asynchronous) client for the Redis
key-value (NoSQL) database server.  Both standard query and subscriber modes
are available.

WHY THIS MODULE?

AnyEvent::Redis2 began as a from-scratch implementation of a Redis module for
AnyEvent.  (When I began writing it, I didn't know such a module already
existed.)  Instead of abandoning it when I realized my oversight, I decided to
contribute it. 

The substantive differences from AnyEvent::Redis are:

- Better documentation
- on_connect() callback 
- Simpler, faster protocol parser (about 30% faster sending large numbers of
  queries, about 10% faster receiving bulk responses)
- Easier-to-read code

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc AnyEvent::Redis
    perldoc AnyEvent::Redis::Subscriber

SOURCE CODE

The official Git repository for AnyEvent::Redis is located at
https://github.com/otterley/AnyEvent-Redis2.

SEE ALSO

Redis <http://redis.io>

LICENSE AND COPYRIGHT

Copyright (C) 2010 Michael S. Fischer

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for AnyEvent-Redis-Lite

0.3     Wed Dec  8 02:25:38 PST 2010  Michael S. Fischer <michael+cpan@dynamine.net>
        - INCOMPATIBLE CHANGE: Changed callback signatures.
        - Fixed and sped up bulk parser dramatically (10x faster now).
        - Added auto-reconnect behavior.
        - Added 'pmessage' event to Subscriber.
        - Documentation updates.
        
0.2     Mon Dec  6 09:21:02 PST 2010  Michael S. Fischer <michael+cpan@dynamine.net>
        Documentation updates.

0.1     Mon Dec  6 00:34:17 PST 2010  Michael S. Fischer <michael+cpan@dynamine.net>
        Initial release.



lib/AnyEvent/Redis2/Protocol.pm
package AnyEvent::Redis2::Protocol;

use common::sense;

=head1 NAME

AnyEvent::Redis2::Protocol - generates and parses Redis messages

=head1 DESCRIPTION

This package should not be directly used.  It provides AnyEvent write and read
types capable of generating Redis requests and parsing replies.

=head1 SEE ALSO

L<AnyEvent::Handle>,
Redis Protocol Specification L<http://code.google.com/p/redis/wiki/ProtocolSpecification>

=cut

sub anyevent_write_type {
    my ($handle, @args) = @_;

    use bytes;
    return "*" . scalar @args . "\015\012" .
          join("\015\012", map { '$' . length($_) . "\015\012" . $_ } @args) .
          "\015\012";
}

sub anyevent_read_type {
    my ($handle, $cb) = @_;
    return sub {
        use bytes;
        $handle->push_read(line => sub {
            my $line = $_[1];
            my $type = substr($line, 0, 1);
            my $value = substr($line, 1);
            if ($type eq '*') {
                # Multi-bulk reply 
                my $remaining = $value;
                if ($remaining == 0) {
                    $cb->([]);
                } elsif ($remaining == -1) {
                    $cb->(undef);
                } else {
                    my $results = [];
                    $handle->push_read(sub {
                        my $need_more_data = 0;
                        do {
                            if ($handle->{rbuf} =~ /^(\$(-?\d+)\015\012)/) {
                                my ($match, $vallen) = ($1, $2);
                                if ($vallen == -1) {
                                    # Delete the bulk header.
                                    substr($handle->{rbuf}, 0, length($match), '');
                                    push @$results, undef;
                                    $remaining--;
                                    unless ($remaining) {
                                        $cb->($results);
                                        return 1;
                                    }
                                } elsif (length $handle->{rbuf} >= (length($match) + $vallen + 2)) {
                                    # OK, we have enough in our buffer.
                                    # Delete the bulk header.
                                    substr($handle->{rbuf}, 0, length($match), '');
                                    push @$results, substr($handle->{rbuf}, 0, $vallen, '');
                                    $remaining--;
                                    # Delete trailing data characters.
                                    substr($handle->{rbuf}, 0, 2, '');
                                    unless ($remaining) {
                                        $cb->($results);
                                        return 1;
                                    }
                                } else {
                                    $need_more_data = 1;
                                }
                            } elsif ($handle->{rbuf} =~ s/^([\+\-:])([^\015\012]*)\015\012//) {
                                my ($type, $value) = ($1, $2);
                                if ($type eq '+' || $type eq ':') {
                                    push @$results, $value;
                                    $remaining--;
                                    unless ($remaining) {
                                        $cb->($results);
                                        return 1;
                                    }
                                } elsif ($type eq '-') {
                                    # Error - abort early
                                    $cb->($value, 1);
                                    return 1;
                                }
                            } else {
                                # Nothing matched - read more...
                                $need_more_data = 1;
                            }
                        } until $need_more_data;
                        return undef; # get more data
                    });
                }
            } elsif ($type eq '+' || $type eq ':') {
                # Single line/integer reply
                $cb->($value);
            } elsif ($type eq '-') {
                # Single-line error reply
                $cb->($value, 1);
            } elsif ($type eq '$') {
                # Bulk reply
                if ($value == -1) {
                    $cb->(undef);
                } else {
                    # We need to read 2 bytes more than the length (stupid 
                    # CRLF framing).  Then we need to discard them.
                    $handle->unshift_read(chunk => $value + 2, sub {
                        my $data = $_[1];
                        $cb->(substr($data, 0, $value));
                    });
                }
            }
            return 1;
        });
        return 1;
    };
}

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai


lib/AnyEvent/Redis2/Subscriber.pm
package AnyEvent::Redis2::Subscriber;

use 5.008;
use common::sense;

use base 'Object::Event';
use Carp qw(croak);
use AnyEvent::Handle;
use AnyEvent::Redis2::Protocol;

=head1 NAME

AnyEvent::Redis2::Subscriber - event-driven Redis subscriber client

=head1 SYNOPSIS

 use AnyEvent::Redis2::Subscriber;
 my $subscriber = AnyEvent::Redis2::Subscriber->subscribe(
     host => 'redis', 
     channel => 'z'
 );
 $subscriber->reg_cb(message => sub {
     my ($subscriber, $message) = @_;
     # ...
 };

=head1 DESCRIPTION

This module is an AnyEvent user; you must use and run a supported event
loop.

AnyEvent::Redis2::Subscriber is an event-driven (asynchronous) subscriber for
the Redis publish/subscribe system.  

The main idea is that other clients (such as L<AnyEvent::Redis2>) publish
messages to a specified channel (or class, in Redis-speak), which are then
received by each client that has subscribed to the channel.

Unlike some other publish/subscribe mechanisms which are producer-consumer
based, Redis' publish/subscribe mechanism is broadcast-based: every subscriber
to a channel receives every message sent to that channel; and messages are not
queued for future delivery if there are no subscribers for that channel.

=head2 Subscribing to a channel

To subscribe to a channel, call the subscribe() method:

 my $subscriber = AnyEvent::Redis2::Subscriber->subscribe(
    host => $host,
    channel => $channel,
    ...
 );

This immediately establishes a connection to the Redis server at the given host
and subscribes to the specified channel. 

Pattern-based subscriptions are also possible.  Instead of specifying a
C<channel>, a C<pattern> may be specified consisting of a glob pattern that
matches channel names.  For example, if C<chan_*> is specified, then the client
will receive every message sent to "chan_1", "chan_2", "chan_zebra", and so
forth while connected to the server:

 my $subscriber = AnyEvent::Redis2::Subscriber->subscribe(
    host => $host,
    pattern => $pattern,
    ...
 );

Optional arguments include:

=over 

=item port => $port

Connect to the Redis server at C<$port>.  If undefined, the default port (6379)
is used.

=item auth => $password

Authenticate to the server with the given password.

=back

=head2 Receiving messages

To receive messages, register a callback for the C<message> event.  When a
message is received, the callback will be fired with the message and the
channel it was received on:

 $subscriber->reg_cb(message => sub {
     my ($channel, $message) = @_;
     print "Received message $message on channel $channel";
 });

If you've subscribed to a channel pattern, you may also register a callback for
the C<pmessage> event.  When a message is received, the callback will be fired
with the message, the channel it was received on, and the pattern:

 $subscriber->reg_cb(pmessage => sub {
     my ($pattern, $channel, $message) = @_;
     print "Received message $message on channel $channel (pattern $pattern)";
 });

=head2 Dealing with errors

To be notified of errors (which can occur at connect time or thereafter), 
register a callback with the C<error> event as follows:

 $subscriber->reg_cb(error => sub {
    my ($errmsg) = @_;
    # ...
 });

B<WARNING:> As with all AnyEvent modules, B<you must not die() in a callback!> 

=head2 Unsubscribing

To unsubscribe, just undef the subscriber handle.  Reusing handles for
subsequent subscriptions (i.e., changing the channel) is not supported.  Even
though the wire protocol technically supports it, trust me, it's better this
way :-).

=cut

sub subscribe {
    my ($class, %args) = @_;
    
    $args{host} or croak "Missing host";
    $args{channel} or $args{pattern} or croak "Missing channel/pattern";
    $args{channel} and $args{pattern} and croak "Both pattern and channel specified";
    $args{port} ||= 6379;

    my $self = $class->SUPER::new(%args);
    bless $self, $class;

    $self->{handle} = AnyEvent::Handle->new(
        connect => [ $self->{host}, $self->{port} ],
        keepalive  => 1,
        no_delay   => 1,
        on_connect => sub { 
            my ($handle, $host, $port) = @_;
            if ($self->{auth}) {
                $self->{handle}->push_write('AnyEvent::Redis2::Protocol', 
                                            AUTH => $self->{auth});
                $self->{handle}->push_read('AnyEvent::Redis2::Protocol' => sub {
                        my ($handle, $errmsg) = @_;
                        if ($errmsg) {
                            $self->{handle}->destroy;
                            $self->event('error', $errmsg);
                        } else {
                            $self->_subscribe;
                        }
                });
            } else {
                $self->_subscribe;
            }
        },
        on_connect_error => sub { 
            $self->event('error', $_[1]);
        },
        on_error => sub { 
            $self->{handle}->destroy; 
            $self->event('error', $_[2]);
        }, 
    );

    return $self;
}

sub _subscribe {
    my $self = shift;
    if ($self->{pattern}) {
        $self->{handle}->push_write('AnyEvent::Redis2::Protocol', 
                                    PSUBSCRIBE => $self->{pattern});
    } else {
        $self->{handle}->push_write('AnyEvent::Redis2::Protocol', 
                                    SUBSCRIBE => $self->{channel});
    }
    $self->{handle}->push_read('AnyEvent::Redis2::Protocol' => sub {
            my ($result, $error) = @_;
            if ($error) { 
                $self->event('error', $result);
            } else {
                if (($self->{channel} && $result->[0] eq 'subscribe' && $result->[1] eq $self->{channel})
                    || ($self->{pattern} && $result->[0] eq 'psubscribe' && $result->[1] eq $self->{pattern})) {
                    $self->{handle}->on_read(sub {
                            $self->{handle}->push_read('AnyEvent::Redis2::Protocol' => sub {
                                    my ($result, $error) = @_;
                                    if ($error) {
                                        $self->event('error', $result);
                                    } else {
                                        if ($result->[0] eq 'message') {
                                            $self->event('message', $result->[1], $result->[2]);
                                        } elsif ($result->[0] eq 'pmessage') {
                                            $self->event('message', $result->[2], $result->[3]);
                                            $self->event('pmessage', $result->[1], $result->[2], $result->[3]);
                                        }
                                    }
                                    1;
                            });
                    });
                } else {
                    $self->event('error', 'Unexpected response to [P]SUBSCRIBE command');
                }
            }
            1;
    });
}

=head1 SEE ALSO

L<AnyEvent::Redis2>, Redis PublishSubscribe
L<http://code.google.com/p/redis/wiki/PublishSubscribe>

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai


lib/AnyEvent/Redis2.pm
package AnyEvent::Redis2;

use 5.008;
use common::sense;

use Carp qw(croak);
use AnyEvent::Handle;
use AnyEvent::Redis2::Protocol;

our $VERSION = 0.3;

=head1 NAME

AnyEvent::Redis2 - an event-driven Redis client

=head1 SYNOPSIS

 use AnyEvent::Redis2;
 my $redis = AnyEvent::Redis2->connect(
     host => 'redis',
     on_connect => $cb, 
     on_connect_error => $errcb,
     on_error => $errcb,
 );

 # Generic query method
 $redis->query('SET', $key, $value, $cb);

 # Autoloaded query method
 $redis->set($key, $value, $cb);

=head1 DESCRIPTION

This module is an AnyEvent user; you must use and run a supported event
loop.

AnyEvent::Redis2 is an event-driven (asynchronous) client for the Redis
key-value (NoSQL) database server.  Every operation is supported, B<except>
subscription to Redis "classes" (i.e. channels), which is supported by
L<AnyEvent::Redis::Subscriber>.  However, this module may be used to publish to
channels.

=head2 Establishing a connection

To connect to the Redis server, use the connect() method:

 my $redis = AnyEvent::Redis2->connect(host => <host>, ...);

The C<host> argument is required.

Optional (but recommended) arguments include:

=over

=item port => $port

Connect to the server on the specified port number.  (If not specified,
the default port will be used.)

=item auth => $password

Authenticate to the server with the given password.

=item on_connect => $cb->($host, $port)

Specifies a callback to be executed upon a successful connection.  The actual
peer host and port number will be passed as arguments to the callback.

=item on_connect_error => $cb->($errmsg)

Specifies a callback to be executed if the connection failed (or authentication
failed).  The error message will be passed to the callback. 

The callback may return an interval value (as a fractional number); if
specified, the client will automatically attempt to reconnect at that interval
until successful.

=item on_error => $cb->($errmsg)

Specifies a callback to be executed if an I/O error occurs (e.g. connection
reset by peer).  The error message will be passed to the callback.

The callback may return an interval value (as a fractional number); if
specified, the client will automatically attempt to reconnect at that interval
until successful.

=back

=cut

sub connect {
    my ($class, %args) = @_;
    
    $args{host} or croak "Missing host";
    $args{port} ||= 6379;

    my $self = { %args };
    bless $self, $class;

    $self->_connect;
    return $self;
}

sub _connect {
    my $self = shift;
    $self->{handle} = AnyEvent::Handle->new(
        connect => [ $self->{host}, $self->{port} ],
        keepalive  => 1,
        no_delay   => 1,
        on_connect => sub { 
            my ($handle, $host, $port) = @_;
            if ($self->{auth}) {
                $self->{handle}->push_write('AnyEvent::Redis2::Protocol', 
                                            AUTH => $self->{auth});
                $self->{handle}->push_read('AnyEvent::Redis2::Protocol' => sub {
                        my ($handle, $ok) = @_;
                        if ($ok) {
                            $self->{on_connect}->($host, $port) 
                                if $self->{on_connect};
                        } else {
                            $self->{handle}->destroy;
                            $self->{on_connect_error}->($_[2])
                                if $self->{on_connect_error};
                        }
                });
            } else {
                $self->{on_connect}->($host, $port) 
                    if $self->{on_connect};
            }
        },
        on_connect_error => sub { 
            if ($self->{on_connect_error}) {
                if (defined(my $interval = $self->{on_connect_error}->($_[1]))) {
                    my $t; $t = AE::timer($interval, 0, sub {
                            $self->_connect;
                            undef $t;
                    });
                }
            }
        },
        on_error   => sub { 
            $self->{handle}->destroy; 
            if ($self->{on_error}) {
                if (defined(my $interval = $self->{on_error}->($_[2]))) {
                    my $t; $t = AE::timer($interval, 0, sub {
                            $self->_connect;
                            undef $t;
                    });
                }
            }
        }, 
    );
}

=head2 Queries and responses

After you have successfully connected to the server, you can issue 
queries to it.  (You'll want to do this in the C<on_connect> callback
handler fired by connect(), above.)  

To issue a query, use the query() method:

  $cv = $redis->query(@args, [ $cb->($data, $error) ]);

The initial list of arguments to query() comprise the actual command.
(See the command reference
L<http://code.google.com/p/redis/wiki/CommandReference> for a list of available
commands.)

The final argument to query() specifies a optional callback (code reference)
that will be fired when a response to the query is received.  It will be called
with the data (either a scalar, or an ARRAY reference for a multi-bulk
response) as the first argument, and a scalar indicating whether the data is an
error message as the second argument.

query() will return an L<AnyEvent> condition variable that can also be used to
retrieve the results via its recv() method (or via cb() if you don't want to
block).  If an error occurs, recv() will die, so be sure to wrap it in an
C<eval>:

  my $cv = $redis->query(@args);
  eval { 
      my $result = $cv->recv; 
      # ...
  };
  if ($@) {
      warn "server error: $@";
      ...
  }

=cut

sub query {
    my ($self, @args) = @_;

    my $cb; $cb = pop @args if ref($_[-1]) eq 'CODE';
    scalar @args or croak "Missing args";

    $args[0] !~ /^P?(?:UN)?SUBSCRIBE/ 
        or croak 'Subscriptions not supported; use AnyEvent::Redis2::Subscriber instead';

    my $cv = AnyEvent->condvar;
    $self->{handle}->push_write('AnyEvent::Redis2::Protocol', @args);
    $self->{handle}->push_read('AnyEvent::Redis2::Protocol' => sub {
            $cb->(@_) if $cb;
            # croak if error occurred; otherwise send data
            $_[1] ? $cv->croak($_[0]) : $cv->send($_[0]);
            1;
    });
    return $cv;
}

=pod

Alternatively, you may invoke Redis commands as methods, e.g.: 

 $cv = $redis->set(key1 => $val1, [ $cb->($result, $error) ]);
 $cv = $redis->lrange('list', 0, 1000, [ $cb->($result, $error) ]);
 $cv = $redis->get(key1, [ $cb->($result, $error) ]);

=cut

sub AUTOLOAD {
    use vars '$AUTOLOAD';
    my $op = $AUTOLOAD;
    $op =~ s/.*:://;
    my ($self, @args) = @_;
    $self->query(uc($op), @args);
}

sub DESTROY {
    # Don't delete this, or AUTOLOAD will be invoked on DESTROY.
}

=head1 WHY NOT AnyEvent::Redis?

AnyEvent::Redis2 began as a from-scratch implementation of a Redis module for
AnyEvent.  (When I began writing it, I didn't know such a module already
existed.)  Instead of abandoning it when I realized my oversight, I decided to
contribute it. 

The substantive differences from AnyEvent::Redis are:

=over

=item Better documentation

=item Automatic reconnection capability

=item on_connect() callback 

=item Simpler, faster protocol parser (about 30% faster sending large numbers
of queries; about 10% faster receiving bulk responses)

=item Easier-to-read code

=item Explicit separation of subscriber functionality (into
AnyEvent::Redis2::Subscriber) 

=back

=head1 SEE ALSO

Redis Command Reference L<http://code.google.com/p/redis/wiki/CommandReference>

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# vim:syn=perl:ts=4:sw=4:et:ai
